### PR TITLE
model/wrapper documentation

### PR DESCRIPTION
**Pull request by hjzccc**

### CODE REVIEW

File: ablator/main/model/wrapper.py [hunk 1]
```diff
@@ -431,6 +637,23 @@ def train_loop(self, smoke_test=False):
     def train(
         self, run_config: RunConfig, smoke_test: bool = False, debug: bool = False
     ) -> TrainMetrics:
+        """
+        Initialize states and train the model.
```
Comment (hieuchi911) [inline ablator/main/model/wrapper.py:636]: In the code, _checkpoint() is called if KeyboardInterrupt error is raised, maybe add this to the description too

File: ablator/main/model/wrapper.py [hunk 2]
```diff
@@ -392,9 +587,20 @@ def eval(self, smoke_test=False):
...
     @property
     def total_steps(self):
+        """
+        The total number of steps for training.
+        """
         return self.epoch_len * self.epochs
 
     def train_loop(self, smoke_test=False):
+        """
+        Train the model in many steps, evaluate the model and log the metrics in between.
```
Comment (hieuchi911) [inline ablator/main/model/wrapper.py]: - I think "log the metrics in between" should be "log the metrics for each iteration".
- Should "metrics" here be clarified as "static metrics like learning rate, along with evaluation and training metrics like loss and mean" so users know what kind of metrics will be logged?

File: ablator/main/model/wrapper.py [hunk 3]
```diff
@@ -362,17 +543,31 @@ def update_status(self):
...
+        """
         # Log step
         if self._is_step(self.log_itr):
             self.metrics.evaluate("train", reset=False)
             self.log_step()
 
     @ty.final
     def eval(self, smoke_test=False):
+        """
+        Evaluate the model then update schedular and save checkpoint if the current iteration is an evaluation step.
```
Comment (hieuchi911) [inline ablator/main/model/wrapper.py]: - typo in `scheduler`
- This method also do early stopping, and check if the iteration is the best iteration and save that (best iteration, best evaluation loss). Should we add it here too?

File: ablator/main/model/wrapper.py [hunk 4]
```diff
@@ -362,17 +543,31 @@ def update_status(self):
         self.train_tqdm.update(1)
 
     def status_message(self) -> str:
+        """
+        Assemble the status message from the current metrics.
```
Comment (hieuchi911) [inline ablator/main/model/wrapper.py]: - I think "status message" is a bit unclear. It calls `self.metrics.to_dict()`, which will returns a dictionary of static and moving average metrics over any set available (e.g train, val) in `TrainerMetrics`

File: ablator/main/model/wrapper.py [hunk 5]
```diff
@@ -362,17 +543,31 @@ def update_status(self):
...
+        --------
+        str
+            The status message.
+        """
         # must return current epoch, iter, losses and metrics
         return " ".join([f"{k}: {v}" for k, v in self.metrics.to_dict().items()])
 
     def log(self):
+        """
+        Log if the current iteration is a logging step.
```
Comment (hieuchi911) [inline ablator/main/model/wrapper.py]: Should we mention that this also evaluate metrics on the train set, so then `log_step` will log all available metrics?

File: ablator/main/model/wrapper.py [hunk 6]
```diff
@@ -347,6 +525,9 @@ def mock_train(
         return mock_model.train(run_config=run_config, smoke_test=True)
 
     def update_status(self):
+        """
+        Update the metrics with current training stats.
```
Comment (hieuchi911) [inline ablator/main/model/wrapper.py]: Should we mention that this update static metrics first, and then all metrics (static and moving average) will be set as description for the tqdm progress?

File: ablator/main/model/wrapper.py [hunk 7]
```diff
@@ -290,6 +427,21 @@ def _model_step(
     def train_step(
         self, batch: Iterable
     ) -> tuple[dict[str, torch.Tensor] | None, dict[str, ty.Any]]:
+        """
+        A single step for training.
```
Comment (hieuchi911) [inline ablator/main/model/wrapper.py:423]: Should we mention that it update learning rate via `scheduler.step()` too?

File: ablator/main/model/wrapper.py [hunk 8]
```diff
@@ -173,13 +278,45 @@ def load_checkpoint(
...
+        """
         if device is None:
             device = self.device
         return butils.iter_to_device(data, device)
 
     def model_step(
         self, model: nn.Module, batch: Iterable
     ) -> tuple[dict[str, torch.Tensor] | None, torch.Tensor | None]:
+        """
+        A single train step for the model.
```
Comment (hieuchi911) [inline ablator/main/model/wrapper.py]: I think this is inference step, aka the model generate predictions, not training step

File: ablator/main/model/wrapper.py [hunk 9]
```diff
@@ -173,13 +278,45 @@ def load_checkpoint(
...
+        model: nn.Module
+            The model to train.
+        batch: Iterable
+            The batch of input data to pass through the model,it could be a list, dict or a single tensor.
+
+        Returns:
+        --------
+        out: dict[str, torch.Tensor] | None
+            The output of the model.
+        loss: torch.Tensor | None
```
Comment (hieuchi911) [inline ablator/main/model/wrapper.py]: loss is not computed in here

File: ablator/main/model/wrapper.py [hunk 10]
```diff
@@ -162,6 +252,21 @@ def reset_optimizer_scheduler(self):
...
     ) -> None:
+        """
+        Loads the checkpoint from the save dict.
+
+        Parameters:
+        -----------
+        save_dict: dict[str, ty.Any]
+            The save dict to load the checkpoint from.
+        model_only: bool
+            Whether to load only the model or include schedular, optimizer and scaler.
```
Comment (hieuchi911) [inline ablator/main/model/wrapper.py]: typo for `scheduler`

File: ablator/main/model/wrapper.py [hunk 11]
```diff
@@ -139,12 +206,35 @@ def create_optimizer(
...
     def reset_optimizer_scheduler(self):
+        """
+        Resets the optimizer and scheduler by recreating them.
+        
+        Returns:
+        --------
+        optimizer: Optimizer
+            The optimizer.
+        scheduler: Scheduler    
+            The scheduler.
```
Comment (hieuchi911) [inline ablator/main/model/wrapper.py]: This method is not returning anything tho

File: ablator/main/model/wrapper.py [hunk 12]
```diff
@@ -334,6 +493,25 @@ def mock_train(
         run_async=True,
         block: bool = True,
     ) -> mp.Process | TrainMetrics:
+        """
+        Mock train the model
```
Comment (hieuchi911) [inline ablator/main/model/wrapper.py]: Should we mention that this will just train the mock model as a smoke test?

File: ablator/main/model/wrapper.py [hunk 13]
```diff
@@ -486,6 +717,27 @@ def apply_loss(
         scaler: torch.cuda.amp.GradScaler,
         scheduler: ty.Optional[Scheduler],
     ) -> float | None:
+        """
+        Calculate the loss and apply the gradients, step the optimizer and scheduler.
```
Comment (hieuchi911) [inline ablator/main/model/wrapper.py]: maybe change "step the optimizer and scheduler" to "call optimizer.step() and scheduler.step()"?

File: ablator/main/model/wrapper.py [hunk 14]
```diff
@@ -542,6 +794,29 @@ def validation_loop(
         subsample: float = 1.0,
         smoke_test: bool = False,
     ) -> dict[str, float]:
+        """
+        Validate the model with the given tag and dataloader.
```
Comment (hieuchi911) [inline ablator/main/model/wrapper.py]: Validate the model on data in `dataloader` (which can either be val dataloader - so tag is `val`, or test dataloader - so tag is `test`)

File: ablator/main/model/wrapper.py [hunk 15]
```diff
@@ -607,6 +946,14 @@ def make_dataloaders(self, run_config: RunConfig) -> None:
         self.test_dataloader = self.make_dataloader_test(run_config)
 
     def checkpoint(self, is_best=False):
+        """
+        Save a checkpoint of the model.
```
Comment (hieuchi911) [inline ablator/main/model/wrapper.py]: Specify how it's saved? i.e directory name (`best` if the trained model achieves the best performance, else `recent`), file name (is named after model class name + current iteration is included in the checkpoint name, etc.)

File: ablator/main/model/wrapper.py [hunk 16]
```diff
@@ -615,6 +962,14 @@ def checkpoint(self, is_best=False):
         )
 
     def save_dict(self) -> dict[str, ty.Any]:
+        """
+        Save the current state of the trainer.
```
Comment (hieuchi911) [inline ablator/main/model/wrapper.py]: "state" includes model parameters, and current states of the optimizer, the scaler, and the scheduler

**Review by hieuchi911 — COMMENTED** [review]
I think the current version is quite short, which might be hard for new users to understand

File: ablator/main/model/wrapper.py (hunk 15 above)
Comment (hjzccc) [reply ablator/main/model/wrapper.py]: I think details can be added in  logger, it just calls the function in logger

File: ablator/main/model/wrapper.py (hunk 2 above)
Comment (MrLYG) [reply ablator/main/model/wrapper.py]: I think "log the metrics for each iteration" is better.

File: ablator/main/model/wrapper.py [hunk 17]
```diff
@@ -542,6 +794,29 @@ def validation_loop(
...
+        dataloader: DataLoader
+            The dataloader to use for validation.
+        metrics: TrainMetrics
+            The metrics to use for validation.
+        tag: ty.Literal["train", "test", "val"]
+            The tag to use for validation. Also see `TrainMetrics` for details.
+        subsample: float
+            The fraction of the dataloader to use for validation.
+        smoke_test: bool
+            Whether it is a smoke test.
```
Comment (MrLYG) [inline ablator/main/model/wrapper.py]:
```suggestion
            Whether to execute this function as a smoke test. If True, only one iteration will be performed, which is useful for quickly checking if the code runs without errors. Default is False.
```

**Review by hieuchi911 — APPROVED** [review]

**Review by MrLYG — APPROVED** [review]

**Review by fostiropoulos — APPROVED** [review]
